### PR TITLE
Coroutines Conversion: rewriting AllReader into coroutines

### PR DESCRIPTION
**Pull request by mikea**

### CODE REVIEW

**Review by jasnell — DISMISSED** [review]

File: src/workerd/api/streams/internal.c++ [hunk 1]
```diff
@@ -96,23 +94,24 @@ private:
...
       runningTotal += amount;
       if (runningTotal >= limit) {
-        return JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
+        throw JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
       }
-      parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
-      return loop();
-    });
+      auto part = kj::heapArray<kj::byte>(amount);
+      memcpy(part.begin(), buffer.begin(), amount);
```
Comment (dom96) [inline src/workerd/api/streams/internal.c++:112]: why the change from `bytes.slice` to `memcpy` here? I'd consider `slice` to be safer.

Comment (mikea) [reply src/workerd/api/streams/internal.c++:112]: Before my change it would allocate several 4K buffers and result would be slices pointing into these buffers. Now we have one buffer. That's why we need to copy its data somewhere.

File: src/workerd/api/streams/internal.c++ [hunk 2]
```diff
@@ -96,23 +94,24 @@ private:
   kj::Vector<kj::Array<kj::byte>> parts;
   uint64_t runningTotal = 0;
 
-  kj::Promise<PartList> loop() {
-    auto bytes = kj::heapArray<kj::byte>(4096);
+  kj::Promise<PartList> readParts() {
+    auto buffer = kj::heapArray<kj::byte>(4096);
```
Comment (jasnell) [inline src/workerd/api/streams/internal.c++:98]: Similar to @kentonv's comment in the other PR, can this be stack allocated instead?

File: src/workerd/api/streams/internal.c++ (hunk 1 above)
Comment (mikea) [reply src/workerd/api/streams/internal.c++]: I managed to remove memcpy anyway. ptal.

File: src/workerd/api/streams/internal.c++ [hunk 3]
```diff
@@ -96,23 +94,23 @@ private:
...
       runningTotal += amount;
       if (runningTotal >= limit) {
-        return JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
+        throw JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
       }
-      parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
-      return loop();
-    });
+
+      parts.add(kj::heapArray<kj::byte>(buffer, amount));
```
Comment (kentonv) [inline src/workerd/api/streams/internal.c++:112]: Hmm, it's too bad this does an extra copy of the data -- although the old code had a different problem, it might have allocated a 4k buffer, read 1 bytes into it, and just kept the rest as wasted space.

Since we're ultimately returning heap buffers, perhaps we should actually allocate our buffer on the heap upfront, and then do `tryRead()` with `minBytes == maxBytes`. This way, the read won't return until either EOF or the buffer is full, so we get the best of both worlds I think.

Comment (mikea) [reply src/workerd/api/streams/internal.c++:112]: yeah, I was thinking the same thing. I wonder how much data is read in practice? One packet? I'll try this out with minBytes == maxBytes tomorrow.

File: src/workerd/api/streams/internal.c++ [hunk 4]
```diff
@@ -96,23 +94,23 @@ private:
...
+  kj::Promise<PartList> readParts() {
+    while (true) {
+      constexpr size_t bufferSize = 4096;
+      auto part = kj::heapArray<kj::byte>(bufferSize);
+      auto amount = co_await input.tryRead(part.begin(), bufferSize, bufferSize);
 
-    return input.tryRead(bytes.begin(), 1, bytes.size())
-        .then([this, bytes = kj::mv(bytes)](size_t amount) mutable
-        -> kj::Promise<PartList> {
       if (amount == 0) {
```
Comment (kentonv) [inline src/workerd/api/streams/internal.c++:103]: You can now change this to `if (amount < bufferSize)`

File: src/workerd/api/streams/internal.c++ [hunk 5]
```diff
@@ -96,23 +94,23 @@ private:
...
-        return KJ_MAP(p, parts) { return p.asPtr(); };
+        co_return KJ_MAP(p, parts) { return p.asPtr(); };
```
Comment (kentonv) [inline src/workerd/api/streams/internal.c++:104]: If `PartList` were `Vector` instead of `Array`, you could avoid the `KJ_MAP` here I think? Save another allocation.

**Comment by mikea** [comment]
btw this requires tweaking upstream test (PR 5828)

File: src/workerd/api/streams/internal.c++ [hunk 6]
```diff
@@ -74,48 +74,48 @@ public:
...
 
+    while (true) {
+      constexpr size_t bufferSize = 4096;
+      auto part = kj::heapArray<kj::byte>(bufferSize);
+      auto amount = co_await input.tryRead(part.begin(), bufferSize, bufferSize);
       runningTotal += amount;
       if (runningTotal >= limit) {
-        return JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
+        throw JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
       }
```
Comment (jasnell) [inline src/workerd/api/streams/internal.c++:106]:
```suggestion
JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
```

**Comment by mikea** [comment]
> There should be no outwardly observable difference after the making the change.

it throws an exception in readAll because of the full buffer now before tee has a chance.

File: src/workerd/api/streams/internal.c++ [hunk 7]
```diff
@@ -61,48 +61,46 @@ public:
...
+      runningTotal += amount;
+      JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
 
-    return input.tryRead(bytes.begin(), 1, bytes.size())
-        .then([this, bytes = kj::mv(bytes)](size_t amount) mutable
-        -> kj::Promise<PartList> {
-      if (amount == 0) {
-        return KJ_MAP(p, parts) { return p.asPtr(); };
+      if (amount > 0) {
+        parts.add(part.slice(0, amount).attach(kj::mv(part)));
```
Comment (kentonv) [inline src/workerd/api/streams/internal.c++:94]: Sorry to keep nitpicking here, but in the common case that `amount == bufferSize`, this `.attach()` is doing a redundant allocation.

Comment (mikea) [reply src/workerd/api/streams/internal.c++]: alright, I got rid of all intermediate copies here.

**Review by jasnell — APPROVED** [review]

**Review by kentonv — APPROVED** [review]
